base/strings/faces.jl

```diff
@@ -0,0 +1,779 @@
+             ifelse(isnothing(b.inverse),       a.inverse,       b.inverse),
+             b.inherit)
+    else
+        merge(merge(a), b)
+    end
+end
+
+"""
+    merge(faces::Face...)
+    merge(faces::Vector{Face})
+
+Merge the properties of `faces` into a single face, with the properties
+of subsequent faces overriding the properties of prior faces.
+"""
+merge(a::Face, b::Face, others::Face...) = merge(merge(a, b), others...)
+merge(faces::Vector{Face}) = foldl(merge, faces, init=Face())
+
+"""
+    merge(face::Face)
+
+Merge any inherited faces of `face` into a self-contained [`Face`](@ref).
+"""
+function merge(face::Face)
+    if isempty(face.inherit)
+        merge(FACES[:default], face)
+    else
+        noinheritface = Face(
+            face.font, face.foreground, face.background, face.weight, face.slant,
+            face.underline, face.strikethrough, face.inverse, Symbol[])
+        inherited = mapfoldl(fname -> get(FACES, fname, Face()), merge,
+                            Iterators.flatten(((:default,),
+                                                Iterators.reverse(face.inherit))))
+        merge(merge(inherited), noinheritface)
+    end
+end
+
+## Getting the combined face from a set of properties ##
+
+"""
+    getface(faces)
+
+Obtain the final merged face from `faces`, an iterator of
+[`Face`](@ref)s, face name `Symbol`s, and lists thereof.
+"""
+function getface(faces)
+    expandface(face::Face) = face
+    expandface(face::Symbol) = get(FACES, face, Face())
+    expandfaces(faces::Vector)::Vector{Face} = expandface.(faces)
+    expandfaces(faces) = (expandface(faces),)
+    foldl(merge,
+          (f::Face for f in Iterators.map(expandfaces, faces) |> Iterators.flatten),
+          init=Face()) |> merge
+end
+
+"""
+    getface(styles::Vector{Pair{Symbol, Any}})
+
+Combine all of the `:face` styles with `getfaces`.
+"""
+function getface(styles::Vector{Pair{Symbol, Any}})
+    faces = (last(prop) for prop in styles if first(prop) === :face)
+    getface(faces)
+end
+
+## Face/StyledString integration ##
+
+"""
+    getface(s::StyledString, i::Integer)
+
+Get the merged [`Face`](@ref) that applies to `s` at index `i`.
+"""
+getface(s::StyledString, i::Integer) =
+    getface(textproperties(s, i))
+
+"""
+    getface(c::StyledChar)
+
+Get the merged [`Face`](@ref) that applies to `c`.
+"""
+getface(c::StyledChar) = getface(c.properties)
+
+"""
+    face!(s::Union{<:StyledString, <:SubString{<:StyledString}},
+          [range::UnitRange{Int},] face::Union{Symbol, Face})
+
+Apply `face` to `s`, along `range` if specified, or the whole of `s`.
+"""
+face!(s::Union{<:StyledString, <:SubString{<:StyledString}},
+      range::UnitRange{Int}, face::Union{Symbol, Face}) =
+          textproperty!(s, range, :face, face)
+
+face!(s::Union{<:StyledString, <:SubString{<:StyledString}},
+      face::Union{Symbol, Face}) =
+          textproperty!(s, firstindex(s):lastindex(s), :face, face)
```

Comment:
You have an example in the PR where you have `face!(str, 5:12, :test)`. 

![image](https://github.com/JuliaLang/julia/assets/1813121/45dc36a5-30ec-4596-82cc-b1703d5bd0c6)

Are the indices into the codepoints or the grapheme indices for the string?

How does it handle printing out an emoji like this:  🤦🏼‍♂️ ? What does the `face!` function do for a string like `str = " 🤦🏼‍♂️  the quick brown fox jumps over the lazy dog"`

This has 5 codepoints:

    🤦 : U+1F926 FACE PALM
    🏼 : U+1F3FC EMOJI MODIFIER FITZPATRICK TYPE-3
    ‍: U+200D ZERO WIDTH JOINER
    ♂ : U+2642 MALE SIGN (Ml)
    ️: U+FE0F VARIATION SELECTOR-16

But some terminals (ones that support it) display it in 1 cell width.

Reply:
It's just Julia string indexing, i.e. codepoints.